### PR TITLE
Fix Plotly.react/Angular missing categories + axis tick labels

### LINKED ISSUE #5318: tick labels missing after interacting with plot from angular-plotly
@archmoj I've had a bit of a play around with a core plotly codepen (https://codepen.io/dmbarry86/pen/BaLKBmR) and a react plotly codepen (https://codepen.io/dmbarry86/pen/VwKaZqa) but am unable to replicate in either sorry. My react codepen does autoresize with window resizing though. I think you were missing style (never have quite understood why this is required TBH).

The only place where I can replicate the issue is the angular-plotly example (https://stackblitz.com/edit/angular-ivy-r3w5lj) shared by @zyzhu. I see in some earlier comments it was mentioned this started in 1.54.5 but when I change versions in that stackblitz it actually seems to be 1.54.6 where the issue starts happening. Not sure if that helps with the debug.

Note that this seems to be a completely different issue to the one originally raised in this ticket which is fixed in 1.58.0 as far as I can tell.

_Originally posted by @dmbarry86 in https://github.com/plotly/plotly.js/issues/4459#issuecomment-739084896_

### PR DESCRIPTION
**Pull request by wbrgss**
Fixes #5318 + adds a test. Many thanks to @alexcjohnson for uncovering this solution and reproducing in CodePen! :slightly_smiling_face:

All Jasmine expectations in the new test should fail on `master`.

Demo:
[Before](https://codepen.io/alexcjohnson/pen/BaLQqby) vs [After](https://codepen.io/wbrgss/pen/WNGpYKG)

***

I also recreated the Angular environment in #5318 and confirmed that this fixes the problem. However, while I can use the build of plotly.js from this PR locally following the [plotly via window module](https://github.com/plotly/angular-plotly.js/blob/master/README.md#plotly-via-window-module) instructions, I unfortunately can't find a way to do the same in the [stackblitz](https://stackblitz.com/edit/angular-ivy-r3w5lj) example — it can't find the Plotly window global on an uploaded `plotly.js` file.

@plotly/plotly_js

### CODE REVIEW

File: test/jasmine/tests/axes_test.js [hunk 1]
```diff
@@ -6896,6 +6896,64 @@ describe('more react tests', function() {
...
+describe('category preservation tests on gd passed to Plotly.react()', function() {
+    var gd;
+
+    beforeEach(function() {
+        gd = createGraphDiv();
+    });
+
+    afterEach(destroyGraphDiv);
+
+    function _hover(gd, opts) {
```
Comment (wbrgss) [inline test/jasmine/tests/axes_test.js:6908]: Borrowed from `hover_label_test.js`

https://github.com/plotly/plotly.js/blob/82da6150acbd56b370c4ccf35733e18de26144bb/test/jasmine/tests/hover_label_test.js#L3712-L3715

File: test/jasmine/tests/axes_test.js [hunk 2]
```diff
@@ -6896,6 +6896,64 @@ describe('more react tests', function() {
...
+                height: 500
+            }
+        };
+
+        Plotly.newPlot(gd, fig)
+        .then(function(gd) {
+            return Plotly.react(gd, fig);
+        })
+        .then(function() {
+            expect(gd._fullLayout.xaxis._categories).toEqual(['a', 'b', 'c', 'd']);
```
Comment (wbrgss) [inline test/jasmine/tests/axes_test.js:6932]: Fails with empty array (`[]`) on `master`

File: test/jasmine/tests/axes_test.js [hunk 3]
```diff
@@ -6896,6 +6896,64 @@ describe('more react tests', function() {
...
+            expect(gd._fullLayout.xaxis._categoriesMap).toEqual({a: 0, b: 1, c: 2, d: 3});
+        })
+        .then(function() {
+            _hover(gd, { xval: fig.data[0].x.indexOf('a') });
+            expect(d3.selectAll('g.axistext').select('text').html()).toEqual('a');
```
Comment (wbrgss) [inline test/jasmine/tests/axes_test.js:6937]: Fails with empty string (`''`) on `master`

**Comment by archmoj** [comment]
Great fix.
Thanks very much for the PR.
:dancer:

**Pull request merged**